File: 2015/cpp/day03.cc
#include <unordered_set>

#include "point.h"
#include "utils.h"

using ::aoc::point::Point;

enum Direction { NORTH, EAST, SOUTH, WEST };

class Santa {
public:
  Santa() : location(0, 0) {}

  void move(const Direction direction) {
    switch (direction) {
    case NORTH:
      location.y++;
      break;
    case EAST:
      location.x++;
      break;
    case SOUTH:
      location.y--;
      break;
    case WEST:
      location.x--;
      break;
    }
  }

  const Point get_location() const { return location; }

private:
  Point location;
};

Direction dir_from_char(const char c) {
  switch (c) {
  case '^':
    return Direction::NORTH;
  case '>':
    return Direction::EAST;
  case 'v':
    return Direction::SOUTH;
  case '<':
    return Direction::WEST;
  default:
    throw std::runtime_error("Unexpected instruction: " + std::string(1, c));
  }
}

int p1(const std::string &instructions) {
  Santa santa;
  std::unordered_set<Point> seen = {santa.get_location()};
  for (const char instruction : instructions) {
    santa.move(dir_from_char(instruction));
    seen.insert(santa.get_location());
  }

  return seen.size();
}

int p2(const std::string &instructions) {
  std::vector<Santa> santas(2);
  std::unordered_set<Point> seen;
  for (const auto &santa : santas) {
    seen.insert(santa.get_location());
  }

  for (std::size_t i = 0; i < instructions.size(); ++i) {
    Santa &santa = santas[i % santas.size()];
    santa.move(dir_from_char(instructions[i]));
    seen.insert(santa.get_location());
  }

  return seen.size();
}

int main() {
  std::string instructions = aoc::utils::get_trimmed_line();

  try {
    std::cout << "Part 1: " << p1(instructions) << std::endl;
    std::cout << "Part 2: " << p2(instructions) << std::endl;
  } catch (const std::exception &exception) {
    std::cerr << exception.what() << std::endl;
    return 1;
  }
}

File: 2015/rust/src/bin/day03.rs
use anyhow::Result;
use rustc_hash::FxHashSet;

use aoc::planar::Direction;
use aoc::utils::get_input;

fn main() -> Result<()> {
    let input = get_input()?;
    let directions: Vec<Direction> = parsing::parse_directions(input.trim())?;

    println!("Part 1: {}", part1(&directions));
    println!("Part 2: {}", part2(&directions));

    Ok(())
}

fn part1(directions: &[Direction]) -> usize {
    houses_visited(directions, 1)
}

fn part2(directions: &[Direction]) -> usize {
    houses_visited(directions, 2)
}

fn houses_visited(directions: &[Direction], santa_count: usize) -> usize {
    if santa_count == 0 {
        return 0;
    }
    let mut visited = FxHashSet::default();
    let mut santas = vec![Santa::default(); santa_count];
    visited.insert(Santa::default());

    for (i, direction) in directions.iter().enumerate() {
        let santa = &mut santas[i % santa_count];
        santa.step(direction);
        visited.insert(*santa);
    }

    visited.len()
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Default)]
struct Santa {
    pub x: i64,
    pub y: i64,
}

impl Santa {
    fn step(&mut self, direction: &Direction) {
        match direction {
            Direction::North => self.y += 1,
            Direction::East => self.x += 1,
            Direction::South => self.y -= 1,
            Direction::West => self.x -= 1,
        }
    }
}

mod parsing {
    use super::Direction;

    use anyhow::{anyhow, Result};

    pub fn parse_directions(s: &str) -> Result<Vec<Direction>> {
        s.chars()
            .map(|c| match c {
                '^' => Ok(Direction::North),
                '>' => Ok(Direction::East),
                'v' => Ok(Direction::South),
                '<' => Ok(Direction::West),
                c => Err(anyhow!("Unrecognized direction {}", c)),
            })
            .collect()
    }
}

#[cfg(test)]
mod tests {
    use super::{houses_visited, parsing::parse_directions};

    #[test]
    fn houses_visited_1() {
        let directions: Vec<_> = parse_directions(">").unwrap();

        assert_eq!(houses_visited(&directions, 1), 2);
    }

    #[test]
    fn houses_visited_2() {
        let directions: Vec<_> = parse_directions("^>v<").unwrap();

        assert_eq!(houses_visited(&directions, 1), 4);
    }

    #[test]
    fn houses_visited_3() {
        let directions: Vec<_> = parse_directions("^v^v^v^v^v").unwrap();

        assert_eq!(houses_visited(&directions, 1), 2);
    }

    #[test]
    fn robo_santa_1() {
        let directions: Vec<_> = parse_directions("^v").unwrap();

        assert_eq!(houses_visited(&directions, 2), 3);
    }

    #[test]
    fn robo_santa_2() {
        let directions: Vec<_> = parse_directions("^>v<").unwrap();

        assert_eq!(houses_visited(&directions, 2), 3);
    }

    #[test]
    fn robo_santa_3() {
        let directions: Vec<_> = parse_directions("^v^v^v^v^v").unwrap();

        assert_eq!(houses_visited(&directions, 2), 11);
    }
}
